src/main/kotlin/org/rust/RsProjectTaskQueueService.kt

```diff
@@ -0,0 +1,223 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust
+
+import com.google.errorprone.annotations.concurrent.GuardedBy
+import com.intellij.openapi.application.ModalityState
+import com.intellij.openapi.components.Service
+import com.intellij.openapi.components.service
+import com.intellij.openapi.progress.*
+import com.intellij.openapi.progress.impl.BackgroundableProcessIndicator
+import com.intellij.openapi.progress.impl.ProgressManagerImpl
+import com.intellij.openapi.project.DumbService
+import com.intellij.openapi.project.Project
+import com.intellij.openapiext.isHeadlessEnvironment
+import com.intellij.openapiext.isUnitTestMode
+import com.intellij.util.PairConsumer
+import com.intellij.util.concurrency.QueueProcessor
+import org.rust.openapiext.DelayedBackgroundableProcessIndicator
+import org.rust.openapiext.checkIsDispatchThread
+import java.util.function.BiConsumer
+
+/**
+ * A common queue for cargo and macro expansion tasks that should be executed sequentially.
+ * Can run any [Task.Backgroundable], but provides additional features for tasks that implement [RsTaskExt].
+ * The most important feature is that newly submitted tasks can cancel ("eat") a currently running task or
+ * tasks in the queue (See [RsTaskExt.eatStrategy]).
+ */
+@Service
+class RsProjectTaskQueueService(project: Project) {
+    private val queue: RsBackgroundTaskQueue = RsBackgroundTaskQueue(project)
+
+    /** Submits a task. A task can implement [RsTaskExt] */
+    fun run(task: Task.Backgroundable) = queue.run(task)
+
+    /** Equivalent to running an empty task with [RsTaskExt.eatStrategy] = [eatStrategy] */
+    fun eatTasks(eatStrategy: RsTaskExt.EatStrategy) = queue.eatTasks(eatStrategy)
+
+    /** @return true if no running or pending tasks */
+    val isEmpty: Boolean get() = queue.isEmpty
+}
+
+val Project.rsTaskQueue: RsProjectTaskQueueService get() = service()
+
+interface RsTaskExt {
+    @JvmDefault
+    val eatStrategy: EatStrategy
+        get() = EatStrategy.INDEPENDENT
+
+    @JvmDefault
+    val progressBarShowDelay: Int
+        get() = 0
+
+    /** If true, the task will not be run (and progress bar will not be shown) until the smart mode */
+    @JvmDefault
+    val waitForSmartMode: Boolean
+        get() = false
+
+    @JvmDefault
+    val runSyncInUnitTests: Boolean
+        get() = false
+
+    /**
+     * Higher position in the enum means higher priority; Newly submitted tasks with higher or equal
+     * priority eats (cancels) other tasks with lower or equal priority if [canBeEaten] == true.
+     * E.g. [CARGO_SYNC] eats [MACROS_UNPROCESSED] but not [MACROS_CLEAR] or itself.
+     * [MACROS_UNPROCESSED] eats itself, [MACROS_FULL] and subsequent.
+     */
+    enum class EatStrategy(val canBeEaten: Boolean = true) {
```

Comment:
Looks like it's not a strategy, it's a task type.
Also, maybe rename `canBeEaten` to `canBeCosumed`?